lib/Service/RecordingService.php

```diff
@@ -0,0 +1,137 @@
+ *
+ */
+
+namespace OCA\Talk\Service;
+
+use InvalidArgumentException;
+use OC\Files\Filesystem;
+use OC\User\NoUserException;
+use OCA\Talk\Config;
+use OCA\Talk\Exceptions\ParticipantNotFoundException;
+use OCA\Talk\Room;
+use OCP\Files\Folder;
+use OCP\Files\IMimeTypeDetector;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotFoundException;
+use OCP\Files\NotPermittedException;
+
+class RecordingService {
+	private IMimeTypeDetector $mimeTypeDetector;
+	private ParticipantService $participantService;
+	private IRootFolder $rootFolder;
+	private Config $config;
+
+	public function __construct(
+		IMimeTypeDetector $mimeTypeDetector,
+		ParticipantService $participantService,
+		IRootFolder $rootFolder,
+		Config $config
+	) {
+		$this->mimeTypeDetector = $mimeTypeDetector;
+		$this->participantService = $participantService;
+		$this->rootFolder = $rootFolder;
+		$this->config = $config;
+	}
+
+	public function store(Room $room, string $owner, array $file): void {
+		$content = $this->getContentFromFileArray($file);
+
+		$recordFileName = $this->sanitizeFileName($file['name']);
+		$this->validateFileFormat($recordFileName, $content);
+
+		try {
+			$this->participantService->getParticipant($room, $owner);
+		} catch (ParticipantNotFoundException $e) {
+			throw new InvalidArgumentException('owner_participant');
+		}
+
+		try {
+			$recordingFolder = $this->getRecordingFolder($owner, $room->getToken());
+			$recordingFolder->newFile($recordFileName, $content);
+		} catch (NoUserException $e) {
+			throw new InvalidArgumentException('owner_invalid');
+		} catch (NotPermittedException $e) {
+			throw new InvalidArgumentException('owner_permission');
+		}
+	}
+
+	public function getContentFromFileArray(array $file): string {
+		if (
+			$file['error'] !== 0 ||
+			!is_uploaded_file($file['tmp_name']) ||
+			Filesystem::isFileBlacklisted($file['tmp_name'])
+		) {
+			throw new InvalidArgumentException('invalid_file');
+		}
+
+		$content = file_get_contents($file['tmp_name']);
+		unlink($file['tmp_name']);
+
+		if (!$content) {
+			throw new InvalidArgumentException('empty_file');
+		}
+		return $content;
+	}
+
+	public function validateFileFormat(string $fileName, $content): void {
+		$mimeType = $this->mimeTypeDetector->detectString($content);
+		$allowed = $this->config->getRecordingAllowedMimes();
+		if (!array_key_exists($mimeType, $allowed)) {
+			throw new InvalidArgumentException('file_mimetype');
+		}
+
+		$extension = strtolower(pathinfo($fileName, PATHINFO_EXTENSION));
+		if (!$extension || !in_array($extension, $allowed[$mimeType])) {
+			throw new InvalidArgumentException('file_extension');
+		}
+	}
+
+	public function sanitizeFileName(string $fileName): string {
+		$recordFileName = escapeshellcmd($fileName);
+		$recordFileName = pathinfo($recordFileName, PATHINFO_BASENAME);
+		if ($recordFileName !== $fileName) {
+			throw new InvalidArgumentException('file_name');
+		}
+		return $recordFileName;
+	}
```

Comment:
This function looks wrong. Not even sure what you are trying to "sanitize" as you only ever return when input = output, so it's more of a validate.

But then again, the `Folder::newFile` will make sure the file name is okay.

Reply:
I will fix the method name.

About the method `Folder::newFile`, I think that this method didn't do this validations.